File: src/gambler/danske_spil.py
from __future__ import annotations

import json
import urllib.parse
import urllib.request
from datetime import datetime, timedelta, timezone
from typing import Any


CONTENT_BASE = "https://content.sb.danskespil.dk/content-service/api/v1/q"

SPORTS: dict[str, dict[str, Any]] = {
    "football": {"drilldown_id": "12", "label": "Football/soccer", "sport_codes": {"FOOTBALL"}},
    "tennis": {"drilldown_id": "854", "label": "Tennis", "sport_codes": {"TENNIS"}},
    "basketball": {"drilldown_id": "465", "label": "Basketball", "sport_codes": {"BASKETBALL"}},
    "formula1": {
        "drilldown_id": "319",
        "label": "Formula 1 / motorsport",
        "sport_codes": {"MOTOR_RACING", "MOTORSPORT"},
        "outright_drilldown_id": "17711",
    },
    "golf": {"drilldown_id": "561", "label": "Golf", "sport_codes": {"GOLF"}},
    "cycling": {"drilldown_id": "660", "label": "Cycling", "sport_codes": {"CYCLING"}},
}

VIRTUAL_MARKERS = ("esoccer", "ebasketball", "efodbold", "ebasket", "esport", "e-sport")


def fetch_json(path: str, params: dict[str, Any]) -> dict[str, Any]:
    query = urllib.parse.urlencode(params, doseq=True)
    request = urllib.request.Request(
        f"{CONTENT_BASE}/{path}?{query}",
        headers={
            "Accept": "application/json",
            "User-Agent": (
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/125.0.0.0 Safari/537.36"
            ),
        },
    )
    with urllib.request.urlopen(request, timeout=25) as response:
        return json.loads(response.read().decode("utf-8"))


def date_boundaries(days: int) -> str:
    now = datetime.now(timezone.utc).replace(minute=0, second=0, microsecond=0)
    return ",".join((now + timedelta(days=offset)).isoformat().replace("+00:00", "Z") for offset in range(days + 1))


def market_kind(market: dict[str, Any]) -> str:
    name = (market.get("name") or "").lower()
    group = (market.get("groupCode") or "").lower()
    outcome_text = " ".join((outcome.get("name") or "") for outcome in market.get("outcomes") or []).lower()
    price_lines = []
    for outcome in market.get("outcomes") or []:
        for price in outcome.get("prices") or []:
            price_lines.extend([price.get("handicapLow"), price.get("handicapHigh")])

    if "outright" in group or ("vinder" in name and "championship" in name):
        return "outright"
    if "special" in name or "special" in group or "specials" in name:
        return "special"
    if "hjorne" in name or "hjørne" in name or "corner" in name or "corners" in group:
        return "corners"
    if "kombination" in name or "kombination" in outcome_text or "combi" in group:
        return "combination"
    if "over" in name or "under" in name or "total" in group or "over_under" in group or " o " in f" {outcome_text} ":
        return "over_under"
    if "mal" in name or "mål" in name or "goal" in group:
        return "goal"
    if "handicap" in name or "handicap" in group or any(value is not None for value in price_lines):
        return "handicap"
    if "begge hold scorer" in name or "both_teams" in group:
        return "both_teams_score"
    if "dobbeltchance" in name or "double_chance" in group:
        return "double_chance"
    if "saet" in name or "sæt" in name or "set" in name or "game" in name:
        return "set_or_game"
    if "halvleg" in name or "half" in name:
        return "half_time"
    if "quarter" in name or "periode" in name:
        return "period_or_quarter"
    if "vinder" in name or "winner" in group or "kampvinder" in name or "match_result" in group:
        return "winner"
    return "other"


def normalize_outcome(outcome: dict[str, Any]) -> dict[str, Any]:
    prices = outcome.get("prices") or []
    price = prices[0] if prices else {}
    return {
        "id": outcome.get("id"),
        "name": outcome.get("name"),
        "type": outcome.get("type"),
        "sub_type": outcome.get("subType"),
        "status": outcome.get("status"),
        "active": outcome.get("active"),
        "displayed": outcome.get("displayed"),
        "decimal_odds": price.get("decimal"),
        "fractional": (
            f"{price.get('numerator')}/{price.get('denominator')}"
            if price.get("numerator") is not None and price.get("denominator") is not None
            else None
        ),
        "handicap_low": price.get("handicapLow"),
        "handicap_high": price.get("handicapHigh"),
    }


def normalize_market(market: dict[str, Any]) -> dict[str, Any]:
    outcomes = [normalize_outcome(outcome) for outcome in market.get("outcomes") or []]
    return {
        "id": market.get("id"),
        "name": market.get("name"),
        "group_code": market.get("groupCode"),
        "kind": market_kind(market),
        "status": market.get("status"),
        "active": market.get("active"),
        "displayed": market.get("displayed"),
        "bet_in_run": market.get("betInRun"),
        "handicap_value": market.get("handicapValue"),
        "minimum_accumulator": market.get("minimumAccumulator"),
        "maximum_accumulator": market.get("maximumAccumulator"),
        "outcome_count": market.get("outcomeCount"),
        "outcomes": outcomes,
    }


def normalize_event(event: dict[str, Any], max_markets: int) -> dict[str, Any]:
    commentary = event.get("commentary") or {}
    facts = []
    for fact in commentary.get("facts") or []:
        participant = next(
            (item for item in commentary.get("participants") or [] if item.get("id") == fact.get("participantId")),
            {},
        )
        facts.append(
            {
                "type": fact.get("type"),
                "value": fact.get("value"),
                "participant": participant.get("name"),
                "role": participant.get("roleCode"),
            }
        )

    return {
        "id": event.get("id"),
        "name": event.get("name"),
        "start_time": event.get("startTime"),
        "started": event.get("started"),
        "live_now": event.get("liveNow"),
        "sort_code": event.get("sortCode"),
        "status": event.get("status"),
        "resulted": event.get("resulted"),
        "settled": event.get("settled"),
        "sport": (event.get("category") or {}).get("name"),
        "sport_code": (event.get("category") or {}).get("code"),
        "class_name": (event.get("class") or {}).get("name"),
        "competition": (event.get("type") or {}).get("name"),
        "competition_drilldown_tag_id": event.get("competitionDrilldownTagId"),
        "external_ids": event.get("externalIds") or [],
        "teams": event.get("teams") or [],
        "market_count": event.get("marketCount"),
        "scoreboard_facts": facts,
        "markets": [normalize_market(market) for market in (event.get("markets") or [])[:max_markets]],
    }


def is_relevant_event(event: dict[str, Any], config: dict[str, Any], include_live: bool) -> bool:
    sport_code = (event.get("category") or {}).get("code")
    if config.get("sport_codes") and sport_code not in config["sport_codes"]:
        return False
    if not include_live and (event.get("started") or event.get("liveNow")):
        return False
    haystack = " ".join(
        str(value or "")
        for value in (event.get("name"), (event.get("class") or {}).get("name"), (event.get("type") or {}).get("name"))
    ).lower()
    return not any(marker in haystack for marker in VIRTUAL_MARKERS)


def fetch_match_events(
    config: dict[str, Any],
    limit: int,
    max_markets: int,
    date_days: int,
    include_live: bool,
) -> list[dict[str, Any]]:
    params: dict[str, Any] = {
        "maxMarkets": max_markets,
        "excludeEventsWithNoMarkets": "false",
        "allowedEventSorts": "MTCH",
        "includeChildMarkets": "true",
        "prioritisePrimaryMarkets": "true",
        "includeCommentary": "true",
        "includeIncidents": "true",
        "includeMedia": "true",
        "drilldownTagIds": config["drilldown_id"],
        "excludeDrilldownTagIds": "20769,22796,22797,22800",
        "useMarketGroupCodeCombis": "true",
        "maxTotalItems": max(limit * 25, 100),
        "maxEventsPerCompetition": min(max(limit * 4, limit), 50),
        "maxCompetitionsPerSportPerBand": 20,
        "maxEventsForNextToGo": 5,
        "startTimeOffsetForNextToGo": 600,
        "lang": "da-DK",
        "channel": "I",
    }
    if not include_live and date_days > 0:
        params["dates"] = date_boundaries(date_days)

    payload = fetch_json("time-band-event-list", params)
    bands = (((payload.get("data") or {}).get("timeBandEvents")) or [])
    events = []
    for band in bands:
        for event in band.get("events") or []:
            if not is_relevant_event(event, config, include_live):
                continue
            events.append(normalize_event(event, max_markets=max_markets))
            if len(events) >= limit:
                return events
    return events


def fetch_outright_events(config: dict[str, Any], limit: int, max_markets: int) -> list[dict[str, Any]]:
    drilldown_id = config.get("outright_drilldown_id")
    if not drilldown_id:
        return []
    payload = fetch_json(
        "event-list",
        {
            "eventSortsIncluded": "TNMT",
            "includeChildMarkets": "true",
            "drilldownTagIds": drilldown_id,
            "lang": "da-DK",
            "channel": "I",
        },
    )
    events = []
    for event in ((payload.get("data") or {}).get("events")) or []:
        if not is_relevant_event(event, config, include_live=True):
            continue
        events.append(normalize_event(event, max_markets=max_markets))
        if len(events) >= limit:
            break
    return events


def summarize_sport(
    sport_key: str,
    limit: int,
    max_markets: int,
    date_days: int = 0,
    include_live: bool = False,
    include_outrights: bool = True,
) -> dict[str, Any]:
    config = SPORTS[sport_key]
    events = fetch_match_events(config, limit, max_markets, date_days, include_live)
    outrights = fetch_outright_events(config, limit, max_markets) if include_outrights else []
    scoped_events = events + outrights
    competitions = sorted({event.get("competition") for event in scoped_events if event.get("competition")})
    market_kinds = sorted(
        {
            market.get("kind")
            for event in scoped_events
            for market in event.get("markets", [])
            if market.get("kind")
        }
    )
    return {
        "sport_key": sport_key,
        "label": config["label"],
        "drilldown_id": config["drilldown_id"],
        "sport_codes": sorted(config.get("sport_codes") or []),
        "observed_at": datetime.now(timezone.utc).isoformat(),
        "date_days": date_days,
        "include_live": include_live,
        "event_count": len(events),
        "outright_count": len(outrights),
        "competitions": competitions[:25],
        "market_kinds": market_kinds,
        "events": events,
        "outrights": outrights,
    }


def scan_sports(
    sports: list[str] | None = None,
    limit: int = 2,
    max_markets: int = 8,
    include_live: bool = False,
) -> dict[str, Any]:
    selected = sorted(SPORTS.keys()) if not sports else sports
    return {
        "source": "content.sb.danskespil.dk content-service",
        "mode": "read_only_anonymous",
        "observed_at": datetime.now(timezone.utc).isoformat(),
        "sports": [
            summarize_sport(
                sport,
                limit=limit,
                max_markets=max_markets,
                include_live=include_live,
            )
            for sport in selected
        ],
    }

File: src/danske_spil.rs
use anyhow::Context;
use chrono::Utc;
use reqwest::Client;
use serde_json::{json, Value};

const CONTENT_BASE: &str = "https://content.sb.danskespil.dk/content-service/api/v1/q";

#[derive(Clone)]
struct SportConfig {
    key: &'static str,
    drilldown_id: &'static str,
    label: &'static str,
    sport_codes: &'static [&'static str],
    outright_drilldown_id: Option<&'static str>,
}

const SPORTS: &[SportConfig] = &[
    SportConfig {
        key: "football",
        drilldown_id: "12",
        label: "Football/soccer",
        sport_codes: &["FOOTBALL"],
        outright_drilldown_id: None,
    },
    SportConfig {
        key: "tennis",
        drilldown_id: "854",
        label: "Tennis",
        sport_codes: &["TENNIS"],
        outright_drilldown_id: None,
    },
    SportConfig {
        key: "basketball",
        drilldown_id: "465",
        label: "Basketball",
        sport_codes: &["BASKETBALL"],
        outright_drilldown_id: None,
    },
    SportConfig {
        key: "motorsports",
        drilldown_id: "319",
        label: "Motorsports",
        sport_codes: &["MOTOR_RACING", "MOTORSPORT"],
        outright_drilldown_id: Some("17711"),
    },
    SportConfig {
        key: "golf",
        drilldown_id: "561",
        label: "Golf",
        sport_codes: &["GOLF"],
        outright_drilldown_id: None,
    },
    SportConfig {
        key: "cycling",
        drilldown_id: "660",
        label: "Cycling",
        sport_codes: &["CYCLING"],
        outright_drilldown_id: None,
    },
];

pub async fn scan_sports(
    limit: usize,
    max_markets: usize,
    include_live: bool,
) -> anyhow::Result<Value> {
    let client = Client::builder()
        .user_agent("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36")
        .build()?;
    let mut sports = Vec::new();
    for sport in SPORTS {
        sports.push(summarize_sport(&client, sport, limit, max_markets, include_live).await?);
    }
    Ok(json!({
        "source": "content.sb.danskespil.dk content-service",
        "mode": "read_only_anonymous",
        "observed_at": Utc::now(),
        "sports": sports
    }))
}

async fn fetch_json(
    client: &Client,
    path: &str,
    params: &[(&str, String)],
) -> anyhow::Result<Value> {
    let url = format!("{CONTENT_BASE}/{path}");
    let payload = client
        .get(url)
        .query(params)
        .send()
        .await
        .context("content-service request failed")?
        .error_for_status()
        .context("content-service returned non-success")?
        .json::<Value>()
        .await
        .context("content-service JSON decode failed")?;
    Ok(payload)
}

async fn summarize_sport(
    client: &Client,
    config: &SportConfig,
    limit: usize,
    max_markets: usize,
    include_live: bool,
) -> anyhow::Result<Value> {
    let events = fetch_match_events(client, config, limit, max_markets, include_live).await?;
    let outrights = if let Some(drilldown_id) = config.outright_drilldown_id {
        fetch_outrights(client, config, drilldown_id, limit, max_markets).await?
    } else {
        Vec::new()
    };
    let event_count = events.len();
    let outright_count = outrights.len();
    let scoped = events.iter().chain(outrights.iter());
    let mut competitions: Vec<String> = scoped
        .clone()
        .filter_map(|event| {
            event
                .get("competition")
                .and_then(Value::as_str)
                .map(str::to_string)
        })
        .collect();
    competitions.sort();
    competitions.dedup();
    let mut market_kinds: Vec<String> = scoped
        .flat_map(|event| {
            event
                .get("markets")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
        })
        .filter_map(|market| {
            market
                .get("kind")
                .and_then(Value::as_str)
                .map(str::to_string)
        })
        .collect();
    market_kinds.sort();
    market_kinds.dedup();

    Ok(json!({
        "sport_key": config.key,
        "label": config.label,
        "drilldown_id": config.drilldown_id,
        "sport_codes": config.sport_codes,
        "observed_at": Utc::now(),
        "date_days": 0,
        "include_live": include_live,
        "event_count": event_count,
        "outright_count": outright_count,
        "competitions": competitions.into_iter().take(25).collect::<Vec<_>>(),
        "market_kinds": market_kinds,
        "events": events,
        "outrights": outrights
    }))
}

async fn fetch_match_events(
    client: &Client,
    config: &SportConfig,
    limit: usize,
    max_markets: usize,
    include_live: bool,
) -> anyhow::Result<Vec<Value>> {
    let params = vec![
        ("maxMarkets", max_markets.to_string()),
        ("excludeEventsWithNoMarkets", "false".to_string()),
        ("allowedEventSorts", "MTCH".to_string()),
        ("includeChildMarkets", "true".to_string()),
        ("prioritisePrimaryMarkets", "true".to_string()),
        ("includeCommentary", "true".to_string()),
        ("includeIncidents", "true".to_string()),
        ("includeMedia", "true".to_string()),
        ("drilldownTagIds", config.drilldown_id.to_string()),
        (
            "excludeDrilldownTagIds",
            "20769,22796,22797,22800".to_string(),
        ),
        ("useMarketGroupCodeCombis", "true".to_string()),
        ("maxTotalItems", usize::max(limit * 25, 100).to_string()),
        (
            "maxEventsPerCompetition",
            usize::min(usize::max(limit * 4, limit), 50).to_string(),
        ),
        ("maxCompetitionsPerSportPerBand", "20".to_string()),
        ("maxEventsForNextToGo", "5".to_string()),
        ("startTimeOffsetForNextToGo", "600".to_string()),
        ("lang", "da-DK".to_string()),
        ("channel", "I".to_string()),
    ];
    let payload = fetch_json(client, "time-band-event-list", &params).await?;
    let mut events = Vec::new();
    for band in payload
        .pointer("/data/timeBandEvents")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        for event in band
            .get("events")
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
        {
            if !is_relevant_event(event, config, include_live) {
                continue;
            }
            events.push(normalize_event(event, max_markets));
            if events.len() >= limit {
                return Ok(events);
            }
        }
    }
    Ok(events)
}

async fn fetch_outrights(
    client: &Client,
    config: &SportConfig,
    drilldown_id: &str,
    limit: usize,
    max_markets: usize,
) -> anyhow::Result<Vec<Value>> {
    let params = vec![
        ("eventSortsIncluded", "TNMT".to_string()),
        ("includeChildMarkets", "true".to_string()),
        ("drilldownTagIds", drilldown_id.to_string()),
        ("lang", "da-DK".to_string()),
        ("channel", "I".to_string()),
    ];
    let payload = fetch_json(client, "event-list", &params).await?;
    let mut events = Vec::new();
    for event in payload
        .pointer("/data/events")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        if !is_relevant_event(event, config, true) {
            continue;
        }
        events.push(normalize_event(event, max_markets));
        if events.len() >= limit {
            break;
        }
    }
    Ok(events)
}

fn is_relevant_event(event: &Value, config: &SportConfig, include_live: bool) -> bool {
    let sport_code = event
        .pointer("/category/code")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if !config.sport_codes.is_empty() && !config.sport_codes.contains(&sport_code) {
        return false;
    }
    if !include_live && (bool_field(event, "started") || bool_field(event, "liveNow")) {
        return false;
    }
    let haystack = format!(
        "{} {} {}",
        text_field(event, "name"),
        event
            .pointer("/class/name")
            .and_then(Value::as_str)
            .unwrap_or_default(),
        event
            .pointer("/type/name")
            .and_then(Value::as_str)
            .unwrap_or_default()
    )
    .to_ascii_lowercase();
    ![
        "esoccer",
        "ebasketball",
        "efodbold",
        "ebasket",
        "esport",
        "e-sport",
    ]
    .iter()
    .any(|marker| haystack.contains(marker))
}

fn normalize_event(event: &Value, max_markets: usize) -> Value {
    let commentary = event.get("commentary").unwrap_or(&Value::Null);
    let participants = commentary
        .get("participants")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let facts: Vec<Value> = commentary
        .get("facts")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .map(|fact| {
            let participant = participants
                .iter()
                .find(|item| item.get("id") == fact.get("participantId"))
                .unwrap_or(&Value::Null);
            json!({
                "type": fact.get("type").cloned().unwrap_or(Value::Null),
                "value": fact.get("value").cloned().unwrap_or(Value::Null),
                "participant": participant.get("name").cloned().unwrap_or(Value::Null),
                "role": participant.get("roleCode").cloned().unwrap_or(Value::Null)
            })
        })
        .collect();

    let markets: Vec<Value> = event
        .get("markets")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .take(max_markets)
        .map(normalize_market)
        .collect();

    json!({
        "id": event.get("id").cloned().unwrap_or(Value::Null),
        "name": event.get("name").cloned().unwrap_or(Value::Null),
        "start_time": event.get("startTime").cloned().unwrap_or(Value::Null),
        "started": event.get("started").cloned().unwrap_or(Value::Null),
        "live_now": event.get("liveNow").cloned().unwrap_or(Value::Null),
        "sort_code": event.get("sortCode").cloned().unwrap_or(Value::Null),
        "status": event.get("status").cloned().unwrap_or(Value::Null),
        "resulted": event.get("resulted").cloned().unwrap_or(Value::Null),
        "settled": event.get("settled").cloned().unwrap_or(Value::Null),
        "sport": event.pointer("/category/name").cloned().unwrap_or(Value::Null),
        "sport_code": event.pointer("/category/code").cloned().unwrap_or(Value::Null),
        "class_name": event.pointer("/class/name").cloned().unwrap_or(Value::Null),
        "competition": event.pointer("/type/name").cloned().unwrap_or(Value::Null),
        "competition_drilldown_tag_id": event.get("competitionDrilldownTagId").cloned().unwrap_or(Value::Null),
        "external_ids": event.get("externalIds").cloned().unwrap_or_else(|| json!([])),
        "teams": event.get("teams").cloned().unwrap_or_else(|| json!([])),
        "market_count": event.get("marketCount").cloned().unwrap_or(Value::Null),
        "scoreboard_facts": facts,
        "markets": markets
    })
}

fn normalize_market(market: &Value) -> Value {
    let outcomes: Vec<Value> = market
        .get("outcomes")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .map(normalize_outcome)
        .collect();
    json!({
        "id": market.get("id").cloned().unwrap_or(Value::Null),
        "name": market.get("name").cloned().unwrap_or(Value::Null),
        "group_code": market.get("groupCode").cloned().unwrap_or(Value::Null),
        "kind": market_kind(market),
        "status": market.get("status").cloned().unwrap_or(Value::Null),
        "active": market.get("active").cloned().unwrap_or(Value::Null),
        "displayed": market.get("displayed").cloned().unwrap_or(Value::Null),
        "bet_in_run": market.get("betInRun").cloned().unwrap_or(Value::Null),
        "handicap_value": market.get("handicapValue").cloned().unwrap_or(Value::Null),
        "minimum_accumulator": market.get("minimumAccumulator").cloned().unwrap_or(Value::Null),
        "maximum_accumulator": market.get("maximumAccumulator").cloned().unwrap_or(Value::Null),
        "outcome_count": market.get("outcomeCount").cloned().unwrap_or(Value::Null),
        "outcomes": outcomes
    })
}

fn normalize_outcome(outcome: &Value) -> Value {
    let price = outcome
        .get("prices")
        .and_then(Value::as_array)
        .and_then(|prices| prices.first())
        .unwrap_or(&Value::Null);
    let fractional = match (price.get("numerator"), price.get("denominator")) {
        (Some(numerator), Some(denominator)) if !numerator.is_null() && !denominator.is_null() => {
            Value::String(format!("{numerator}/{denominator}"))
        }
        _ => Value::Null,
    };
    json!({
        "id": outcome.get("id").cloned().unwrap_or(Value::Null),
        "name": outcome.get("name").cloned().unwrap_or(Value::Null),
        "type": outcome.get("type").cloned().unwrap_or(Value::Null),
        "sub_type": outcome.get("subType").cloned().unwrap_or(Value::Null),
        "status": outcome.get("status").cloned().unwrap_or(Value::Null),
        "active": outcome.get("active").cloned().unwrap_or(Value::Null),
        "displayed": outcome.get("displayed").cloned().unwrap_or(Value::Null),
        "decimal_odds": price.get("decimal").cloned().unwrap_or(Value::Null),
        "fractional": fractional,
        "handicap_low": price.get("handicapLow").cloned().unwrap_or(Value::Null),
        "handicap_high": price.get("handicapHigh").cloned().unwrap_or(Value::Null)
    })
}

fn market_kind(market: &Value) -> &'static str {
    let name = text_field(market, "name").to_ascii_lowercase();
    let group = text_field(market, "groupCode").to_ascii_lowercase();
    let outcome_text = market
        .get("outcomes")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|outcome| outcome.get("name").and_then(Value::as_str))
        .collect::<Vec<_>>()
        .join(" ")
        .to_ascii_lowercase();
    let has_handicap = market
        .get("outcomes")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .flat_map(|outcome| {
            outcome
                .get("prices")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
        })
        .any(|price| price.get("handicapLow").is_some() || price.get("handicapHigh").is_some());

    if group.contains("outright") || (name.contains("vinder") && name.contains("championship")) {
        "outright"
    } else if name.contains("special") || group.contains("special") {
        "special"
    } else if name.contains("hjorne")
        || name.contains("hjørne")
        || name.contains("corner")
        || group.contains("corners")
    {
        "corners"
    } else if name.contains("kombination")
        || outcome_text.contains("kombination")
        || group.contains("combi")
    {
        "combination"
    } else if name.contains("over")
        || name.contains("under")
        || group.contains("total")
        || group.contains("over_under")
    {
        "over_under"
    } else if name.contains("mål") || name.contains("mal") || group.contains("goal") {
        "goal"
    } else if name.contains("handicap") || group.contains("handicap") || has_handicap {
        "handicap"
    } else if name.contains("begge hold scorer") || group.contains("both_teams") {
        "both_teams_score"
    } else if name.contains("dobbeltchance") || group.contains("double_chance") {
        "double_chance"
    } else if name.contains("sæt")
        || name.contains("saet")
        || name.contains("set")
        || name.contains("game")
    {
        "set_or_game"
    } else if name.contains("halvleg") || name.contains("half") {
        "half_time"
    } else if name.contains("quarter") || name.contains("periode") {
        "period_or_quarter"
    } else if name.contains("vinder")
        || group.contains("winner")
        || name.contains("kampvinder")
        || group.contains("match_result")
    {
        "winner"
    } else {
        "other"
    }
}

fn text_field<'a>(value: &'a Value, field: &str) -> &'a str {
    value.get(field).and_then(Value::as_str).unwrap_or_default()
}

fn bool_field(value: &Value, field: &str) -> bool {
    value.get(field).and_then(Value::as_bool).unwrap_or(false)
}
